apps/src/code-studio/pd/foorm/editor/components/FoormEntityEditorPreview.jsx

```diff
@@ -0,0 +1,66 @@
+// Preview panel for Foorm editor.
+import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import FontAwesome from '@cdo/apps/templates/FontAwesome';
+import Foorm from '../../Foorm';
+import {connect} from 'react-redux';
+
+const styles = {
+  previewBox: {
+    border: '1px solid #eee'
+  },
+  errorMessage: {
+    fontWeight: 'bold',
+    padding: '1em'
+  }
+};
+
+class FoormEntityEditorPreview extends Component {
+  static propTypes = {
+    previewQuestions: PropTypes.object,
+    forceRerenderKey: PropTypes.number,
+    surveyData: PropTypes.object,
+    errorMessages: PropTypes.array,
+
+    // Populated by Redux
+    hasJSONError: PropTypes.bool
+  };
+
+  renderErrorMessages() {
+    return this.props.errorMessages.map((errorMessage, i) => {
+      return (
+        <div style={styles.errorMessage} key={i}>
+          <FontAwesome icon="exclamation-triangle" /> {errorMessage}
+        </div>
+      );
+    });
+  }
+
+  render() {
+    return (
+      <div className="foorm-preview">
+        <div style={styles.previewBox}>
+          {this.renderErrorMessages()}
+          {this.props.previewQuestions &&
+            !this.props.hasJSONError &&
+            !(this.props.errorMessages && this.props.errorMessages.length) && (
```

Comment:
i only recently learned that javascript supports optional chaining (haven't seen any examples in our codebase other than a recent PR), but this would be a good use for it: `this.props.errorMessages?.length`

Reply:
TIL (among other things)! Thanks for the tip.

Reply:
Bah maybe not:
```
ERROR in ./src/code-studio/pd/foorm/editor/components/FoormEntityEditorPreview.jsx
Module build failed (from ./node_modules/babel-loader/lib/index.js):
SyntaxError: /Users/benjaminbrooks/code-dot-org/apps/src/code-studio/pd/foorm/editor/components/FoormEntityEditorPreview.jsx: Support for the experimental syntax 'optionalChaining' isn't currently enabled
```

Reply:
weird! i'm gonna have to look up the PR that used it.